Fix the presentation submission in the oid4vp-rs e2e test.

Ryanmtate opened the pull request:
As part of this fix, upgrading `ssi` dependency to `0.8.1`;

- [x] Upgrade ssi version to 0.8.1; (use ssi sub-crates)
- [x] Construct a `jwt_vp` token;
- [x] Return a `jwt_vp` token instead of the `jtw_vc` token in the Presentation Submission;
- [x]  Validate the Presentation Submission against the Presentation Definition;
- [x] Validate the VP Token flow, see: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.5 and https://identity.foundation/presentation-exchange/spec/v2.0.0/#input-evaluation;

NOTE: Saving the following for another PR:
- Handle JWT encoded authorization request.

Ryanmtate committed:
wip: use ssi 0.8.1, debugging e2e test using did resolver

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/core/authorization_request/verification/did.rs:
```diff
@@ -6,17 +6,29 @@ use crate::core::{
-    resolver: &dyn DIDResolver,
-) -> Result<()> {
-    let (headers_b64, _, _) = ssi::jws::split_jws(&request_jwt)?;
+    resolver: &VerificationMethodDIDResolver<impl DIDResolver, M>,
```
Not sure if this is the correct resolver type to use here.

Possibly change to `impl JWKResolver` type?

cobward replied:
Yep, you can remove the associated type and use:

```rust
    pub async fn new(
        vm: String,
        signer: Arc<dyn RequestSigner + Send + Sync>,
        resolver: &VerificationMethodDIDResolver<
            impl DIDResolver,
            impl MaybeJwkVerificationMethod
                + VerificationMethodSet
                + TryFrom<GenericVerificationMethod, Error = InvalidVerificationMethod>,
        >,
```

Ryanmtate commented on src/core/authorization_request/verification/did.rs:
```diff
@@ -64,11 +78,12 @@ pub async fn verify_with_resolver(
         }
     }
 
-    let jwk = resolve_key(&kid, resolver)
+    let jwk = resolver
+        .fetch_public_jwk(Some(&kid))
         .await
-        .context("unable to resolve verification method from 'kid' header")?;
+        .context("unable to fetch JWK from 'kid' header")?;
 
-    let _: Json = ssi::jwt::decode_verify(&request_jwt, &jwk)
+    let _: Json = ssi::claims::jwt::decode_verify(&request_jwt, &jwk)
```
Given the upgrade to `0.8.1`, the `resolve_key` method is removed.

I changed the argument type to `VerificationMethodDIDResolver` since it provides implementations of `DIDResolver` and `JWKResolver`.

That said, this is not working at the moment, resulting in the following error: 

```
thread 'w3c_vc_did_client_direct_post' panicked at tests/e2e.rs:49:58:
called `Result::unwrap()` on an `Err` value: unable to validate authorization request

Caused by:
    0: unable to validate Authorization Request
    1: unable to fetch JWK from 'kid' header
    2: internal error: DID resolution failed: DID method `key` not supported
```

My sense is that `fetch_public_key` isn't the right method to resolve the JWK.

Should I use a DIDResolver `Output` and parse the public key to create an `JWK` instance instead?

cobward replied:
I think that is just because you are using DID-JWK instead of DID-Key in the test:
![image](https://github.com/user-attachments/assets/234a849b-d61f-4474-bbc3-e8648d2f99bb)

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -94,9 +104,9 @@ pub struct PresentationSubmission {
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
 pub struct DescriptorMap {
     pub id: String,
-    pub format: String, // TODO should be enum of supported formats
+    pub format: VerifiableFormat,
```
Test cases work for this replacement.

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -84,6 +84,16 @@ pub enum ConstraintsLimitDisclosure {
     Preferred,
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum VerifiableFormat {
+    #[serde(rename = "jwt_vc_json")]
+    JwtVcJson,
+    #[serde(rename = "jwt_vp_json")]
+    JwtVpJson,
+    #[serde(rename = "ldp_vc")]
+    LdpVc,
+}
+
```
Are there any other formats we'd like to include here?

Should `jwt_vp` and `jwt_vc` be separate options in addition to these?

Ryanmtate replied:
NOTE: I can also stash these changes and apply to another PR if preferable.

cobward replied:
This is fine, but please add `Other(String)` to capture other variants.

Ryanmtate replied:
Using the `ClaimFormat` and `ClaimFormatDesgination` types added in #8. Will add the `Other(String)` type to allow for ad hoc types.

cobward replied:
Could you also reuse this enum as the key for the maps [here](https://github.com/spruceid/oid4vp-rs/blob/main/src/core/metadata/parameters/wallet.rs#L137)
and [here](https://github.com/spruceid/oid4vp-rs/blob/main/src/core/metadata/parameters/verifier.rs#L8).

Ryanmtate replied:
Commit [1a2af67](https://github.com/spruceid/oid4vp-rs/pull/8/commits/1a2af6719c47bd8827f43504c0048480e8d273db) adds a designation for `jwt_vc_json` and `jwt_vp_json` with the wallet supported types claim format.

Since technically, its a presentation format more so than a claim format, it might make sense to break this into a separate type, or rename `ClaimFormat` to something more generic, possibly `Format` and `FormatDesignation`.

I'm open to suggestions on the naming and whether it makes sense for it to belong where I put it in the above commit.

Ryanmtate commented on src/verifier/client.rs:
```diff
@@ -43,17 +43,18 @@ impl DIDClient {
         vm: String,
         signer: Arc<dyn RequestSigner + Send + Sync>,
-        resolver: &dyn DIDResolver,
+        resolver: impl JWKResolver,
```
This may be incorrect resolver type. 

I'm assuming the verification method type needs to resolve into a vm-specific key type, not necessarily a JWK.

Ryanmtate commented on src/verifier/client.rs:
```diff
@@ -43,17 +43,18 @@ impl DIDClient {
     pub async fn new(
         vm: String,
         signer: Arc<dyn RequestSigner + Send + Sync>,
-        resolver: &dyn DIDResolver,
+        resolver: impl JWKResolver,
     ) -> Result<Self> {
         let (id, _f) = vm.rsplit_once('#').context(format!(
             "expected a DID verification method, received '{vm}'"
         ))?;
 
-        let key = ssi::did_resolve::resolve_key(&vm, resolver)
+        let key = resolver
+            .fetch_public_jwk(Some(&vm))
             .await
             .context("unable to resolve key from verification method")?;
 
-        if &key != signer.jwk() {
+        if &*key != signer.jwk() {
```
Used the `VerificationMethodDIDResolver` type since checking the JWK is performed, but not sure this is the correct key resolution process, using `fetch_public_jwk` method above.

Throws the following error:

```
thread 'w3c_vc_did_client_direct_post' panicked at tests/jwt_vc.rs:53:10:
called `Result::unwrap()` on an `Err` value: unable to resolve key from verification method

Caused by:
    internal error: DID resolution failed: DID method `key` not supported
```

cobward replied:
As above, I think this is correct, you just need to pass in the correct DID resolver.

Ryanmtate replied:
Perfect, that was the issue! Thanks for the tip!

Ryanmtate committed:
wip: use VerificationMethodDIDResolver for DIDClient constructor

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on tests/jwt_vc.rs:
```diff
@@ -36,11 +40,14 @@ pub async fn wallet_verifier() -> (JwtVcWallet, Arc<Verifier>) {
     );
+
+    let resolver = VerificationMethodDIDResolver::new(DIDJWK);
+
     let client = Arc::new(
-        oid4vp::verifier::client::DIDClient::new(
+        oid4vp::verifier::client::DIDClient::new::<AnyMethod>(
             verifier_did_vm.clone(),
             signer.clone(),
-            DIDKey.to_resolver(),
+            &resolver,
```
I changed the `DIDClient` constructor to take `VerificationMethodDIDResolver`. Not sure if this is the correct verification method resolver or setup.

Ryanmtate committed:
remove unused imports

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
wip: debugging did resolver jwk not found in e2e flow

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
feat/improve-presentation-exchange-support-in-oid4vp-rs

add implementation methods for Presentation Definition.

WIP: Need to continue work for Presentation Submission and the rest
of the structs used in the presentation exchange flow.

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Co-authored-by: Todd Showalter <todd.showalter@spruce.id>

Ryanmtate committed:
add getter methods for presentation definition member fields

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
fix broken links in documentation

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented:
Merge #9 before this.

Ryanmtate committed:
update presentation submission implementation

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
fix test cases. todo: update test cases to use newly created interface for presentation exchange

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
update json schema validator to use anyhow result type

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
merge presentation exchange updates

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

cobward commented on Cargo.toml:
```diff
@@ -25,7 +25,7 @@ serde_cbor = "0.11.2"
 serde_qs = "0.12.0"
 serde_urlencoded = "0.7.1"
-ssi = "0.7.0"
+ssi = "0.8.1"
```
Would be great if you could also refactor to just use the specific sub-crates that we need in this repo to contract the dependency tree, but feel free to do this once everything is implemented and working.

cobward commented on src/verifier/client.rs:
```diff
@@ -40,20 +47,26 @@ pub struct DIDClient {
-    pub async fn new(
+    pub async fn new<M>(
         vm: String,
         signer: Arc<dyn RequestSigner + Send + Sync>,
-        resolver: &dyn DIDResolver,
-    ) -> Result<Self> {
+        resolver: &VerificationMethodDIDResolver<impl DIDResolver, M>,
```
Could you remove the associated type using the same trick as above?

cobward reviewed (commented):
Great progress so far!

Ryanmtate committed:
wip: use latest implementation changes, update tests

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
fix verification method did resolver tests

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
wip: remove unused imports

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/json_schema_validation.rs:
```diff
@@ -0,0 +1,204 @@
+        }
+
+        if let Some(pattern) = &self.pattern {
+            // Note: In a real implementation, you'd use a regex library here
```
Replace with regex library for pattern matching.

Ryanmtate committed:
add ClaimFormat type

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add regex support for string pattern matching

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
revert uuid presentation definition id type to string

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
fix: ensure negation of regex pattern match for error

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add 'other' variante to claim format type

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
remove commented out code

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
merge presentation exchange updates

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented:
NOTE: This branch includes updates to be merged in #9.

Ryanmtate committed:
use ssi sub-crates instead of main ssi dependency

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add jwt_vc_json and jwt_vp_json claim formats

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/core/object/mod.rs:
```diff
@@ -13,7 +13,7 @@ pub struct UntypedObject(pub(crate) Map<String, Json>);
 /// A strongly typed parameter that can represent metadata entries or request parameters.
 pub trait TypedParameter:
-    TryFrom<Json, Error = anyhow::Error> + Into<Json> + Clone + std::fmt::Debug
+    TryFrom<Json, Error = anyhow::Error> + TryInto<Json> + Clone + std::fmt::Debug
```
Add TryInto::Error type restriction to use anyhow::Error.

Ryanmtate replied:
Adding the error type will cause conflicts with the default From<T> infallible error type. Leaving off the TryInto::Error type.

Ryanmtate committed:
wip: construct verifiable presentation for e2e test

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
wip: perform validation on presentation submission

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/holder/verifiable_presentation_builder.rs:
Considering yanking this. Currently, it is not being used.

Ryanmtate commented on src/core/authorization_request/parameters.rs:
```diff
@@ -193,7 +193,28 @@ impl TryFrom<Json> for ClientMetadataUri {
+    }
+}
+
+impl Nonce {
+    pub fn random() -> Self {
+        // NOTE: consider replacing with rng rand crate.
+        Self(uuid::Uuid::new_v4().to_string())
+    }
+}
```
Replace with a better nonce generator?

cobward replied:
```suggestion
impl Nonce {
    pub fn random(length: u8) -> Self {
        Self(Alphanumeric.sample_string(&mut rand::thread_rng(), length))
    }
}
```

Ryanmtate committed:
ensure range exclusive values are checked; fix inclusive range values

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -1,80 +1,772 @@
+        auth_response: &AuthorizationResponse,
+    ) -> Result<()> {
+        match auth_response {
+            AuthorizationResponse::Jwt(jwt) => {
+                bail!("Authorization Response Presentation Definition Validation Not Implemented.")
+            }
```
TODO: Handle Jwt encoded authorized response.

Ryanmtate replied:
Should we resolve this now or in another PR?

cobward replied:
Another PR

Ryanmtate committed:
ensure enum values are parsed in alpha descending order

This is a fix for a bug where ClaimFormat::JwtVc was being parsed
when ClaimFormat::JwtVp should have been instead. The fix is to
order the enum fields in alphabetical descending order, such that
VP comes BEFORE VC, and so on, for the other formats.

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
wip: verify authorized response presentation submission

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/json_schema_validation.rs:
```diff
@@ -0,0 +1,344 @@
+/// - [https://json-schema.org/understanding-json-schema](https://json-schema.org/understanding-json-schema)
+///
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct SchemaValidator {
```
Add a [`const`](https://json-schema.org/understanding-json-schema/reference/const) value for equality check or restricted to a certain value. 

Add [`enum`](https://json-schema.org/understanding-json-schema/reference/enum) type to ensure `one-of` predicate for inclusion in a set.

Ryanmtate replied:
Ensure the validator matches the definitions specified here: https://json-schema.org/draft/2020-12/json-schema-validation

Ryanmtate replied:
This is handled in #9 and has been merged into this branch.

Ryanmtate committed:
ensure json schema validator adheres to the specification

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
ensure json schema validator adheres to the specification

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add unit tests for schema validator

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
use serde default value for constraints field if not found during deserialization

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
merge schema validator changes

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -1,80 +1,846 @@
+/// for an example schema.
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum ClaimFormat {
```
There is a bug when serializing and deserializing the `ClaimFormat` type, resulting in `JwtVcJson` being deserialized into `JwtVpJson`.

This occurs in the e2e test. Re-arranging the enum field order does change the behavior, but does not fix the bug.

Marking this problem as it will have significant side effects and is difficult to detect. So it should be tracked for a bugfix.

Ryanmtate replied:
This has been resolved. Removing the `untagged` seems to do the trick.

Ryanmtate committed:
remove unused imports

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
Merge branch 'main' into fix/fix-the-presentation-submission-in-the-oid4vp-rs-e2e-test

Ryanmtate committed:
remove unsed imports in test files

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
update vp token

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
rebase with main

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
resolve merge conflict; remove schema validator and replace with jsonschema

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
remove unused dependencies

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -277,6 +319,71 @@ impl PresentationDefinition {
+
+                let jwt = response.vp_token().0.clone();
+
+                // TODO: Verify the JWT.
+                // let jws = CompactJWSString::from_string(jwt.clone()).context("Invalid JWT.")?;
+                // let resolver: VerificationMethodDIDResolver<DIDKey, AnyJwkMethod> =
+                //     VerificationMethodDIDResolver::new(DIDKey);
+                // let params = VerificationParameters::from_resolver(resolver);
+
+                // if let Err(e) = jws.verify(params).await {
+                //     bail!("JWT Verification Failed: {:?}", e)
+                // }
```
TODO: Verify the JWT

Ryanmtate replied:
commit [1398cad](https://github.com/spruceid/oid4vp-rs/pull/8/commits/1398cadffd13c2a03fc69fc8514656a7830c3f73) resolves

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -386,6 +493,109 @@ impl InputDescriptor {
+
+                    // If a filter is available with a valid schema, handle the field validation.
+                    if let Some(Ok(schema_validator)) = validator.as_ref() {
+                        // TODO: possible trace a warning if a field is not valid.
```
currently using `debug!` to trace filter validation errors. Possibly unnecessary.

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -386,6 +493,109 @@ impl InputDescriptor {
+                        }
+                    }
+
+                    // TODO: Check limit disclosure of data requested. Do not provide more data
+                    // than is necessary to satisfy the constraints.
```
Remove, is handled above starting on line 547.

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -575,11 +789,47 @@ impl ConstraintsField {
+    ///
+    /// If the filter schema is invalid, this will also return None.
+    ///
+    /// NOTE: Errors are not handled if the filter schema is invalid,
+    /// instead the method will return None on an invalid filter.
```
Remove comment. Validation errors are surfaced.

Ryanmtate commented on tests/e2e.rs:
```diff
@@ -41,47 +57,54 @@ async fn w3c_vc_did_client_direct_post() {
+            // of the descriptor map matching the input descriptor id.
+            DescriptorMap::new(
+                descriptor.id().clone(),
+                ClaimFormatDesignation::JwtVc,
```
Add the following code comment:

```
// NOTE: Since the input descriptor may support several different claim format types. This value should not be 
// hardcoded in production code, but should be selected from available formats in the presentation definition 
// input descriptor.
```

Ryanmtate commented on tests/jwt_vp.rs:
This file will create and save a new VP Token.

Ryanmtate commented on tests/jwt_vp.rs:
```diff
@@ -0,0 +1,117 @@
+use ssi_dids::DIDKey;
+use ssi_jwk::JWK;
+
+#[tokio::test]
```
Determine if this test should be ignored so it does not overwrite existing vp token.

NOTE: the vp token expires after 1 hour, so any test checking expiration will need to generate a valid token.

cobward commented on tests/jwt_vp.rs:
```diff
@@ -0,0 +1,117 @@
+    println!("JWT: {:?}", jwt);
+
+    // Save the JWT to the file system to be used in other tests `examples/vp.jwt`
+    std::fs::write("tests/examples/vp.jwt", jwt.as_str())?;
```
Could we just have this function return it rather than using the FS? I have bad experiences with tests relying on local state :)

Ryanmtate committed:
add rand crate and provide random nonce method using Rng trait

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
remove todos and update comments, use JWKResolver instead of VerificationMethodDIDResolver

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
fix clippy warnings

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
verify jwt in validate_authorization_response presentation definition method

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented:
Once this receives approval, I will squash these commits.

cobward commented on Cargo.toml:
```diff
@@ -9,27 +9,28 @@ repository = "https://github.com/spruceid/oidc4vp-rs/"
+default = []
 reqwest = ["dep:reqwest"]
 p256 = ["dep:p256"]
+rand = ["dep:rand"]
```
Unnecessary feature overhead, `rand` is already in the dependency tree through the ssi crates. And `reqwest` is too as of adding `jsonschema`, so we can remove that feature too. `p256` should be optional, but with the ssi refactor it looks like that is always imported through `ssi-dids`, even with --no-default-features. I think that is a bug, but we can remove it as a feature for now.

Ryanmtate replied:
Did we want the ability to pass in a reference to an existing RNG? e.g.

```rust
impl Nonce {
    pub fn random(rng: &mut impl rand::Rng, len: u8) -> Self {
        use rand::distributions::Alphanumeric;

        Self((0..len).map(|_| rng.sample(Alphanumeric) as char).collect())
    }
}
```

I don't know if this is a common case where an RNG is already initialized and could be passed in.

Ryanmtate replied:
I didn't see `p256` as a re-export from `ssi-dids` for `0.2.0`: https://docs.rs/ssi-dids/latest/ssi_dids/#reexports

I also don't see `rand` as a re-export from `jsonschema`: https://docs.rs/jsonschema/latest/jsonschema/#reexports

How are these libraries used from dependencies, if they are not re-exported?

cobward replied:
```
$ cargo tree -i rand
rand v0.8.5
├── num-bigint-dig v0.8.4
│   └── rsa v0.6.1
│       └── ssi-jwk v0.2.1
...
```

```
$ cargo tree -i reqwest@0.12.5
reqwest v0.12.5
└── jsonschema v0.18.0
```

```
$ cargo tree -i p256
p256 v0.13.2
├── did-method-key v0.2.2
│   [dev-dependencies]
│   └── oid4vp v0.1.0 (/Users/jacob/Documents/git/oidc4vp-rs)
│       [dev-dependencies]
│       └── oid4vp v0.1.0 (/Users/jacob/Documents/git/oidc4vp-rs) (*)
├── did-method-key v0.3.0
│   └── ssi-dids v0.2.0
...
```

cobward replied:
> How are these libraries used from dependencies, if they are not re-exported?

We don't need to use the crates from those dependencies, there's just no point gating them behind features if they're already in the dependency tree.

cobward commented on Cargo.toml:
```diff
@@ -46,3 +47,9 @@ uuid = { version = "1.2", features = ["v4", "serde", "js"] }
 
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
 uuid = { version = "1.2", features = ["v4", "serde"] }
+
+# For Development Purpose, patch the ssi* crates to use the local versions
+# [patch.crates-io]
+# ssi-claims = { path = "../ssi/crates/claims" }
+# ssi-dids = { path = "../ssi/crates/dids" }
+# ssi-jwk = { path = "../ssi/crates/jwk" }
```
My preference is to keep this out of the Cargo.toml, as it makes assumptions about the developer's directory structure.

Ryanmtate replied:
Do you have another preference for patching libs? Or just remove the commented code from the Cargo.toml and let the developer add it as necessary?

cobward replied:
Yep the latter.

cobward commented on src/core/metadata/parameters/wallet.rs:
```diff
@@ -134,7 +137,7 @@ impl From<RequestObjectSigningAlgValuesSupported> for Json {
 
 // TODO: Better types
```
This TODO comment can be removed I think

cobward commented on src/holder/verifiable_presentation_builder.rs:
@timothee-haudebourg @sbihel  what do you think about adding something like this to `ssi` instead of here?

cobward commented on src/verifier/mod.rs:
```diff
@@ -119,12 +119,12 @@ impl Verifier {
-        F: FnOnce(Session, AuthorizationResponse) -> Pin<Box<Fut>>,
-        Fut: Future<Output = Outcome>,
+        F: FnOnce(Session, AuthorizationResponse) -> Fut,
+        Fut: Future<Output = Result<Outcome>>,
```
This is intentionally not `Result<Outcome>`. If there is an error, it should be captured in the `Outcome::Failure` variant.

cobward replied:
This still needs to be reverted.

cobward commented on src/verifier/client.rs:
```diff
@@ -43,17 +44,20 @@ impl DIDClient {
-        resolver: &dyn DIDResolver,
+        resolver: impl JWKResolver,
     ) -> Result<Self> {
+        // use ssi_jwk::JWKResolver;
```
```suggestion
```

cobward commented on tests/e2e.rs:
```diff
@@ -8,80 +10,112 @@ use oid4vp::{
+            Constraints::new()
+                .add_constraint(
+                    ConstraintsField::new(
+                        "$.vp.verifiableCredential[0].vc.credentialSubject.id".into(),
```
```suggestion
                        "$.credentialSubject.id".into(),
```

cobward commented on tests/e2e.rs:
Could you make something more like (or an exact replica of) this example from the specification?
https://openid.net/specs/openid-4-verifiable-presentations-1_0-20.html#name-w3c-verifiable-credentials

Ryanmtate replied:
Yeah, do we want to change the `vc.jwt` example payload? I based the Presentation Definition loosely on the example vc, but it is non-normative.

Ryanmtate replied:
Here's the payload of the example vc currently used:

```json
{
  "iss": "did:key:zDnaeeex9MAVbhoWeDcbbGZdzM1zxqZqpC387jWoLhUr1BdST",
  "nbf": 1704067200,
  "jti": "2c282eba-d147-422c-9ce5-991f199800c5",
  "sub": "did:key:zDnaefqT1BrGGsJEZGwAiueouqMh6MqsZhaL1md5hkHgtfzb2",
  "vc": {
    "@context": "https://www.w3.org/2018/credentials/v1",
    "id": "2c282eba-d147-422c-9ce5-991f199800c5",
    "type": "VerifiableCredential",
    "credentialSubject": {
      "id": "did:key:zDnaefqT1BrGGsJEZGwAiueouqMh6MqsZhaL1md5hkHgtfzb2"
    },
    "issuer": "did:key:zDnaeeex9MAVbhoWeDcbbGZdzM1zxqZqpC387jWoLhUr1BdST",
    "issuanceDate": "2024-01-01T00:00:00+00:00"
  }
}
```

Ryanmtate replied:
Commit https://github.com/spruceid/oid4vp-rs/pull/8/commits/9997486535fbf5ffe1c7f4968f280aef4f9f9337 updates the path to be top-level. But I am also wondering myself if this is correct.

It would seem the top level should be the `Verifiable Submission`, which I would think would include a `jwt_vp_json` located at the `vp` verifiable presentation field.

In that case, shouldn't the top-level `jwt_vp_json` be the `$.vp` field? and not at the presentation submission level? Or is the presentation submission assumed to be a `jwt_vp_json` response payload.

I went with your suggestion, because I tried to track down some example cases in the specification, but didn't see anything definitive.

That said, the descriptor map, when constructing the presentation should know the relative path, and it should match one of the constraint field paths from the input descriptor when constructed.

cobward replied:
We should follow the profile defined in OID4VP:

1. Verifier requests a `jwt_vc_json` with some contraints on that VC.
2. Holder generates a `jwt_vp_json` which is a presentation of that VC, including the transaction's nonce in the VP.
3. Holder presents the VP, identifying where the VC can be found within the VP.

cobward replied:
Just to point out: OID4VP is a framework specification. The semantics are not well defined, and it's down to the "profile" to define how a request is interpreted, and what kind of response to be built.

cobward commented on tests/e2e.rs:
```diff
@@ -8,80 +10,112 @@ use oid4vp::{
+        .set_format((|| {
+            let mut map = ClaimFormatMap::new();
+            map.insert(
+                ClaimFormatDesignation::JwtVp,
```
```suggestion
                ClaimFormatDesignation::JwtVcJson,
```

cobward commented on tests/e2e.rs:
```diff
@@ -8,80 +10,112 @@ use oid4vp::{
-
-    ))
-    .unwrap();
+    let descriptor_map = parsed_presentation_definition
+        .parsed()
+        .input_descriptors()
+        .iter()
+        .map(|descriptor| {
+            // NOTE: the input descriptor constraint field path is relative to the path
+            // of the descriptor map matching the input descriptor id.
+            DescriptorMap::new(
+                descriptor.id().to_string(),
+                // NOTE: Since the input descriptor may support several different claim format types. This value should not be
+                // hardcoded in production code, but should be selected from available formats in the presentation definition
+                // input descriptor.
+                //
+                // In practice, this format will be determined by the VDC collection's credential format.
+                ClaimFormatDesignation::JwtVc,
+                "$".into(),
+            )
+        })
+        .collect();
```
This doesn't make sense to me, shouldn't we presenting something like:
```json
    {
      "id": "did-key-id",
      "path": "$",
      "format": "jwt_vp_json",
      "path_nested": {
        "path": "$.vp.verifiableCredential[0]",
        "format": "jwt_vc_json"
      }
    }
```

Ryanmtate replied:
I've updated the example to more closely resemble your suggestion, see: 

https://github.com/spruceid/oid4vp-rs/pull/8/files#diff-62403d73a2340f1d6ac792a3d62161b538ffefa05bc7365f96351854116d63b1R99-R124

NOTE: This is still using the existing `vc.jwt` example credential.

cobward commented on tests/e2e.rs:
```diff
@@ -8,80 +10,112 @@ use oid4vp::{
+
+    let presentation_submission = PresentationSubmission::new(
+        uuid::Uuid::new_v4(),
+        parsed_presentation_definition.parsed().id().clone(),
```
Must the presentation definition id match the presentation submission id? If they _can_ be different then I think it should be done so here in the example.

Ryanmtate replied:
Presentation Submission has its own unique ID, however, the `definition_id` field within the presentation submission must match the originating presentation definition ID in the authorization request.

Ryanmtate committed:
Update tests/e2e.rs

Co-authored-by: Jacob <jacob.ward@spruceid.com>

Ryanmtate committed:
update descriptor map nested path in e2e example

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
remove dependency patches

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add paths to example for input descriptor constraints field

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
use top level json path for jwt_vp_json

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
Update src/verifier/client.rs

Co-authored-by: Jacob <jacob.ward@spruceid.com>

Ryanmtate committed:
rebase

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
revert validation function async signature to use boxed pin future

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/presentation_exchange.rs:
```diff
@@ -277,6 +320,71 @@ impl PresentationDefinition {
+            AuthorizationResponse::Unencoded(response) => {
+                let presentation_submission = response.presentation_submission().parsed();
+
+                let jwt = response.vp_token().0.clone();
+
+                let jws = CompactJWSString::from_string(jwt.clone()).context("Invalid JWT.")?;
+                let resolver: VerificationMethodDIDResolver<DIDKey, AnyJwkMethod> =
+                    VerificationMethodDIDResolver::new(DIDKey);
+
+                let params = VerificationParameters::from_resolver(resolver);
+
+                if let Err(e) = jws.verify(params).await {
+                    bail!("JWT Verification Failed: {:?}", e)
+                }
```
NOTE: The e2e test is currently failing due to `MissingPublicKey`.

```
unexpected status: Complete(Error { cause: JWT Verification Failed: MissingPublicKey })
```

I'm assuming I am incorrectly signing the verifiable presentation claim? Is this the correct flow?

https://github.com/spruceid/oid4vp-rs/pull/8/files#diff-6d6c7d16f3d6119451ac6b8c431591a112676ff96645d7a6cba9055aa85a7d2aR95-R100

Ryanmtate replied:
Looking at the companion test wallet, it seems I should be using the `Presentation::generate_jwt` from `ssi_vc` to generate the signed JWT.

This brings up another question: should we bump the version of ssi-vc `0.2.*` -> `0.3.*`?

It looks like `Presentation` has become a trait and not a struct type.

For now, I will use the `0.2.1` version and see if I can get the e2e test working correctly.

Ryanmtate replied:
I've removed the JWT encoding and the verification from the e2e test. Using the unencoded flow. Will open a new PR for the JWT processing.

Ryanmtate commented on src/holder/verifiable_presentation_builder.rs:
Should we yank this? It is dead code, and the `holder` module is not used.

I think the helper methods should go directly on the `VerifiablePresentation` `ssi-jwt` claims type.

cobward replied:
Yep let's yank then.

Ryanmtate committed:
add helper methods

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
debug: jwt claim signing does not include public key

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
make request signer methods return a result

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
refactor presentation exchange file into smaller modules

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
fix clippy warnings

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on tests/jwt_vp.rs:
```diff
@@ -0,0 +1,51 @@
+        ssi_claims::jwt::decode_unverified(include_str!("examples/vc.jwt"))?;
+
+    let verifiable_presentation =
+        VerifiablePresentationBuilder::from_options(VerifiablePresentationBuilderOptions {
+            issuer: holder_did.clone(),
+            subject: holder_did.clone(),
+            audience: verifier_did.clone(),
+            expiration_secs: 3600,
+            credentials: vec![verifiable_credential],
+            nonce: "random_nonce".into(),
+        });
```
Decided to keep the verifiable presentation builder and provide a helper method. We may still want to move this closer to the VerifiablePresentation struct itself, but at least this will provide a helper method for reuse.

Ryanmtate committed:
Merge branch 'main' into fix/fix-the-presentation-submission-in-the-oid4vp-rs-e2e-test

Ryanmtate committed:
fix outcome error cause

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add credential format and add requested fields helper method to input descriptor

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

cobward commented on src/core/credential_format/mod.rs:
```diff
@@ -15,3 +19,248 @@ pub struct JwtVc;
+pub enum ClaimFormat {
+    #[serde(rename = "jwt")]
+    Jwt {
+        // The algorithm used to sign the JWT.
```
Please use `///` for all of these comments

cobward commented on src/core/credential_format/mod.rs:
```diff
@@ -15,3 +19,248 @@ pub struct JwtVc;
 impl CredentialFormat for JwtVc {
     const ID: &'static str = "jwt_vc";
 }
+
```
If the old trait isn't used anymore you can delete it.

Ryanmtate replied:
These are now removed.

cobward commented on src/core/input_descriptor.rs:
```diff
@@ -0,0 +1,513 @@
+    }
+
+    /// Validate the input descriptor against the verifiable presentation and the descriptor map.
+    pub fn validate_verifiable_presentation(
+        &self,
+        verifiable_presentation: &VerifiablePresentation,
+        descriptor_map: &DescriptorMap,
+    ) -> Result<()> {
```
Could you add some positive and negative unit tests for this method?

Ryanmtate replied:
Working on these tests. I am covering the submission requirement groups, which were previously not handled in the input descriptors. Using the `tests/presentation-exchange` submodule for example presentation definitions and submissions.

cobward commented on src/core/presentation_definition.rs:
```diff
@@ -0,0 +1,185 @@
+            AuthorizationResponse::Jwt(_jwt) => {
+                // TODO: Handle JWT Encoded authorization response.
+
+                bail!("Authorization Response Presentation Definition Validation Not Implemented.")
```
```suggestion
                bail!("Authorization Response Presentation Definition validation not implemented.")
```

cobward commented on src/core/presentation_definition.rs:
```diff
@@ -0,0 +1,185 @@
+                                    &verifiable_presentation,
+                                    descriptor,
+                                )
+                                .context("Input Descriptor Validation Failed.")?;
```
```suggestion
                                .context("Input Descriptor validation failed.")?;
```

cobward commented on src/core/presentation_definition.rs:
```diff
@@ -0,0 +1,185 @@
+                let jwt = response.vp_token().0.clone();
+
+                let verifiable_presentation: VerifiablePresentation =
+                    ssi_claims::jwt::decode_unverified(&jwt)?;
```
Rather than assuming that the vp_token is a jwt_vp_json, could you do a match on the expected format first?

Ryanmtate replied:
I removed the jwt encoding of the vp_token, and followed the docs more closely for a base64 encoded payload, which now handled if multiple VPs are included in the VP Token.


> REQUIRED. JSON String or JSON object that MUST contain a single Verifiable Presentation or an array of JSON Strings and JSON objects each of them containing a Verifiable Presentations. Each Verifiable Presentation MUST be represented as a JSON string (that is a Base64url encoded value) or a JSON object depending on a format as defined in Appendix A of [[OpenID.VCI](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#OpenID.VCI)]. When a single Verifiable Presentation is returned, the array syntax MUST NOT be used. If Appendix A of [[OpenID.VCI](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#OpenID.VCI)] defines a rule for encoding the respective Credential format in the Credential Response, this rules MUST also be followed when encoding Credentials of this format in the vp_token response parameter. Otherwise, this specification does not require any additional encoding when a Credential format is already represented as a JSON object or a JSON string.

see: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2

cobward commented on src/lib.rs:
```diff
@@ -1,5 +1,7 @@
 pub mod core;
-pub mod presentation_exchange;
+pub mod holder;
```
```suggestion
```

Ryanmtate replied:
I did end up keeping the `holder` module, but I think it is temporary, and would like to move these builder methods closer to the `VerifiablePresentation` struct in the future.

Ryanmtate replied:
Keeping this for now until the functionality is handled closer to the `jwt::VerifiablePresentation`.

cobward commented on src/core/metadata/mod.rs:
```diff
@@ -89,6 +117,18 @@ impl WalletMetadata {
+        };
+
+        Ok(algs.0)
+    }
```
No need to add this method, this should be done via dereference to UntypedObject.

Ryanmtate replied:
This method has been removed.

Ryanmtate committed:
Update src/core/presentation_definition.rs

Co-authored-by: Jacob <jacob.ward@spruceid.com>

Ryanmtate committed:
Update src/core/presentation_definition.rs

Co-authored-by: Jacob <jacob.ward@spruceid.com>

Ryanmtate committed:
wip: add notes on required fields parsing

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
update vp token base64 encoding and check for multiple vp payloads

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
move validation to auth response impl instead of presentation definition impl

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
remove feature gated non-optional deps

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
remove cfg features

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
wip: handle groups in presentation definition, input descriptor tests

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
remove cfg feature tags

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on src/tests.rs:
```diff
@@ -0,0 +1,192 @@
+    }
+}
+
+#[test]
+fn test_input_descriptor_validation() -> Result<()> {
+    // Include the `input_descriptors_example.json` file in the `examples` directory.
+    let input_descriptors = include_str!(
+        "../tests/presentation-exchange/test/presentation-definition/input_descriptors_example.json"
+    );
```
FYI: Submission Requirement groups is currently not handled in the validation flow.

Working on the `group` submission requirements in the validation flow to make this test work. Using the examples from the test examples git submodule.

Ryanmtate replied:
Submission requirement groups are now handled.

Ryanmtate committed:
remove extraneous metadata helper methods; use UntypedObject for dereferencing

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add submission requirement check for presentation validation

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate commented on tests/e2e.rs:
```diff
@@ -1,87 +1,146 @@
+    #[cfg(feature = "rand")]
+    let nonce = Nonce::random(&mut rand::thread_rng());
+
+    #[cfg(not(feature = "rand"))]
```
TODO: Remove unused cfg feature.

Ryanmtate commented on tests/e2e.rs:
```diff
@@ -1,87 +1,146 @@
 
     let client_metadata = UntypedObject::default();
 
+    #[cfg(feature = "rand")]
```
TODO: removed unused cfg feature.

Ryanmtate commented on tests/jwt_vc.rs:
```diff
@@ -152,10 +159,15 @@ impl AsyncHttpClient for MockHttpClient {
+                            },
+                            Err(e) => Outcome::Error {
+                                cause: e.to_string(),
+                            },
```
I've moved the `validate` method to the `auth_response` implementation instead of a method of the presentation definition. I think that makes more sense.

Ryanmtate commented on src/core/presentation_submission.rs:
```diff
@@ -0,0 +1,151 @@
+    ///
+    /// See: [https://identity.foundation/presentation-exchange/spec/v2.0.0/#processing-of-submission-entries](https://identity.foundation/presentation-exchange/spec/v2.0.0/#processing-of-submission-entries)
+    ///
+    /// Errors:
+    /// - The id of the nested path must be the same as the parent id.
```
TODO: Remove this error message, we're no longer throwing an error in this method.

Ryanmtate committed:
add validate method to vp token; ensure submission requirement all rule is enforced.

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
add vp token validate unencoded method. fix minor todos.

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

cobward commented on Cargo.toml:
```diff
@@ -40,7 +40,7 @@ x509-cert = "0.2.4"
 tokio = { version = "1.32.0", features = ["macros"] }
 did-method-key = "0.2"
-oid4vp = { path = ".", features = ["p256"] }
+oid4vp = { path = "." }
```
```suggestion
```

cobward commented on src/core/credential_format/mod.rs:
```diff
@@ -1,17 +1,291 @@
+    },
+    #[serde(rename = "mso_mdoc")]
+    MsoMDoc(serde_json::Value),
+    Other(serde_json::Value),
```
I think this `Other` case needs `#[serde(untagged)]` otherwise it will only capture:
```json
{
  ...
  "Other": { ... }
  ...
}
```

Could you add a little unit test for deserializing and the `designation` fn for this case?

cobward commented on src/core/credential_format/mod.rs:
```diff
@@ -1,17 +1,291 @@
+    /// Other claim format designations not covered by the above.
+    ///
+    /// The value of this variant is the name of the claim format designation.
+    Other(String),
```
As above I think this needs a `#[serde(untagged)]`. Please add a test.

cobward commented on src/core/credential_format/mod.rs:
```diff
@@ -1,17 +1,291 @@
+    // TODO: Add additional credential types here. Fallback to a string for any other credential type.
+    /// Other credential types not covered by the above.
+    Other(String),
+}
```
We will never be able to enumerate all credential types, so I think a string is the most reasonable representation.

cobward commented on src/core/credential_format/mod.rs:
```diff
@@ -1,17 +1,291 @@
+// TODO: Does the `isomdl` crate provide this constant value, or another crate?
+const ORG_ISO_18013_5_1_MDL: &str = "org.iso.18013.5.1.mDL";
+
+/// A Json object of claim formats.
+pub type ClaimFormatMap = HashMap<ClaimFormatDesignation, ClaimFormatPayload>;
```
Why is this needed in addition to `ClaimFormat`? Isn't `ClaimFormatMap` equivalent to `Vec<ClaimFormat>`?

Ryanmtate replied:
The presentation exchange presentation definition test suite test case `format_example` shows `format` property as a map. I tried with the `Vec<ClaimFormat>` but encountered errors in the test case.

```json
{
  "presentation_definition": {
    "id": "32f54163-7166-48f1-93d8-ff217bdb0653",
    "input_descriptors": [],
    "format": {
      "jwt": {
        "alg": ["EdDSA", "ES256K", "ES384"]
      },
      "jwt_vc": {
        "alg": ["ES256K", "ES384"]
      },
      "jwt_vp": {
        "alg": ["EdDSA", "ES256K"]
      },
      "ldp_vc": {
        "proof_type": [
          "JsonWebSignature2020",
          "Ed25519Signature2018",
          "EcdsaSecp256k1Signature2019",
          "RsaSignature2018"
        ]
      },
      "ldp_vp": {
        "proof_type": ["Ed25519Signature2018"]
      },
      "ldp": {
        "proof_type": ["RsaSignature2018"]
      }
    }
  }
}

```

cobward commented on src/core/input_descriptor.rs:
```diff
@@ -0,0 +1,610 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    purpose: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    format: Option<ClaimFormatMap>,
```
Using [KeyValueMap](https://docs.rs/serde_with/latest/serde_with/struct.KeyValueMap.html), you can use `Vec<ClaimFormat>` instead of `ClaimFormatMap` for better typed values.

cobward commented on src/core/input_descriptor.rs:
```diff
@@ -0,0 +1,610 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    format: Option<ClaimFormatMap>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    group: Option<Vec<GroupId>>,
```
Is it important to distinguish between `None` and an empty vec/map? If not:
```suggestion
    #[serde(default, skip_serializing_if = "ClaimFormatMap::is_empty")]
    format: ClaimFormatMap,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    group: Vec<GroupId>,
```

cobward commented on src/core/input_descriptor.rs:
```diff
@@ -0,0 +1,610 @@
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
+pub struct Constraints {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    fields: Option<Vec<ConstraintsField>>,
```
As above, may not be important to distinguish between `None` and an empty vec.

cobward commented on src/core/input_descriptor.rs:
```diff
@@ -0,0 +1,610 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    optional: Option<bool>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    intent_to_retain: Option<bool>,
```
As above, may be better to use the default boolean value rather than `None`

cobward commented on src/core/presentation_definition.rs:
```diff
@@ -0,0 +1,382 @@
+    name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    purpose: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    format: Option<ClaimFormatMap>,
+}
```
As above, can use `Vec<ClaimFormat>`

Ryanmtate replied:
Removed the `Option` and but maintained `ClaimFormatMap` to pass test suite.

cobward commented on src/core/response/mod.rs:
```diff
@@ -35,6 +38,51 @@ impl AuthorizationResponse {
+                    bail!("Presentation Definition ID does not match the Presentation Submission.")
+                }
+
+                // Validate the VP Token as an unencoded payload.
+                response.vp_token().validate_unencoded(
+                    presentation_definition,
+                    presentation_submission.descriptor_map(),
+                )?;
```
This assumes that the vp_token is a W3C VCDM Verifiable Presentation. That is only one possibility. The VP token could be a base64-encoded `DocumentResponse` from ISO 18013-5/-7, an SD-JWT+KB, or a completely proprietary data format.

Validation is incredibly complex and dependent on many different contextual factors, so for now, I suggest removing this function in favour of individual methods implementing specific checks that we know we can always perform (such as the definition ID matching).

Ryanmtate replied:
Validation methods have been removed in commit [0e0f9](https://github.com/spruceid/oid4vp-rs/pull/8/commits/0e0f985da4e5e0827efedf86509a424f25cc4816).

cobward commented on src/core/response/parameters.rs:
```diff
@@ -46,28 +52,96 @@ impl From<VpToken> for Json {
 }
 
+impl VpToken {
+    /// Parse the VP Token as a JSON object.
+    ///
+    /// This will attempt to decode the token as base64, and if that fails, it
+    /// will attempt to parse the token as a JSON object.
+    ///
+    /// See: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2
+    ///
+    /// If you want to check for decode errors, use [VpToken::decode_base64].
+    pub fn parse(&self) -> Result<Json, Error> {
+        match self.decode_base64() {
+            Ok(decoded) => Ok(decoded),
+            Err(_) => Ok(serde_json::from_str(&self.0)?),
+        }
+    }
+
+    /// Return the Verifiable Presentation Token as a JSON object from a base64
+    /// encoded string.
+    pub fn decode_base64(&self) -> Result<Json, Error> {
+        let decoded = BASE64_STANDARD.decode(&self.0)?;
+        Ok(serde_json::from_slice(&decoded)?)
+    }
```
These methods are inaccurate:

> JSON String or JSON object that MUST contain a single Verifiable Presentation or an array of JSON Strings and JSON objects each of them containing a Verifiable Presentations. Each Verifiable Presentation MUST be represented as a JSON string (that is a Base64url encoded value) or a JSON object depending on a format as defined in Annex E of [[OpenID.VCI](https://openid.net/specs/openid-4-verifiable-presentations-1_0-20.html#OpenID.VCI)]

We have a bug in our definition of `VpToken`. It should be something like:

```rust
enum VpToken {
    Single(String),
    SingleAsMap(Map<String, Value>),
    Many(Vec<VpToken>),
}
```

Ryanmtate replied:
I have a TODO in the code to update the VpToken. Maybe we can do this in another PR.

cobward commented on src/core/response/parameters.rs:
```diff
@@ -46,28 +52,96 @@ impl From<VpToken> for Json {
+    /// the requirements of the presentation definition.
+    ///
+    ///
+    pub fn validate_unencoded(
```
As above, this is making an assumption that the VpToken is a VCDM VP. There's also some confusion here about encoded as a JWT vs unencoded. Encoded as a JWT refers to the authorization response (see [direct_post.jwt](https://openid.net/specs/openid-4-verifiable-presentations-1_0-20.html#name-response-mode-direct_postjw)), not the VP token. The VP token is never encoded directly as a JWT. It's either a String, a JSON object, or an array of VP tokens.

Ryanmtate replied:
This method has been removed.

cobward commented on src/holder/verifiable_presentation_builder.rs:
I think we said this should be removed from this library and added to SSI?

Ryanmtate replied:
Okay to remove in another PR once added to SSI?

Ryanmtate committed:
Update Cargo.toml

Co-authored-by: Jacob <jacob.ward@spruceid.com>

Ryanmtate committed:
fix other claim format serde

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
use Vec::is_empty versus Option::is_none for various serialization fields

This commit also removes validation logic from vp token response struct.

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

Ryanmtate committed:
revert to use of ClaimFormatMap to pass presentation defintion test suite

Signed-off-by: Ryan Tate <ryan.michael.tate@gmail.com>

timothee-haudebourg commented on src/holder/verifiable_presentation_builder.rs:
```diff
@@ -0,0 +1,164 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use serde::{Deserialize, Serialize};
+use ssi_claims::jwt::{VerifiableCredential, VerifiablePresentation};
```
I think there is a confusion here, probably because of the lack of documentation in ssi I'll admit. Those two types are wrappers around the [JWT registered claims `vc` and `vp`](https://www.iana.org/assignments/jwt/jwt.xhtml) to be used with `ssi_claims::jwt::ClaimSet`.

If you want to represent VCs and VPs, the appropriate types are defined in `ssi_claims::vc`.

timothee-haudebourg commented on src/holder/verifiable_presentation_builder.rs:
```diff
@@ -0,0 +1,164 @@
+
+#[derive(Debug, Clone)]
+pub struct VerifiablePresentationBuilderOptions {
+    pub issuer: DIDURLBuf,
```
Is it a requirement of OID4VP that we use DIDs? I know this is not a requirement of VPs.

timothee-haudebourg commented on src/holder/verifiable_presentation_builder.rs:
```diff
@@ -0,0 +1,164 @@
+    ///
+    /// This will set the issuance date to the current time and the expiration
+    /// date to the expiration secs from the issuance date.
+    pub fn from_options(options: VerifiablePresentationBuilderOptions) -> VerifiablePresentation {
```
This sounds like a re-implementation of VCDM1.0 JWT VP encoding. This is already implemented in `ssi` [here](https://github.com/spruceid/ssi/blob/bfdf22d14ac96bef7e8fc0c88fa9980198da699d/crates/claims/crates/vc/src/v1/jwt/mod.rs#L88).

```rust
// Create the presentation.
let presentation: ssi::claims::vc::v1::JsonPresentation = ...;

// Encode the JWT claims according to VCDM1.0 JWT VP format.
let jwt_claims = presentation.to_jwt_claims();

// Sign the claims into a JWT.
let jwt = jwt_claims.sign(signer).await.unwrap();
```

timothee-haudebourg commented on src/core/credential_format/mod.rs:
```diff
@@ -1,17 +1,305 @@
+    #[serde(rename = "jwt_vp")]
+    JwtVp,
+    #[serde(rename = "jwt_vc_json")]
+    JwtVcJson,
```
What is this format?

timothee-haudebourg reviewed (commented):
Adding a late review of this to clarify some stuff about `ssi`.
